### ISSUE TITLE
'cursorline' priority: low priority if only bg is set

### ISSUE THREAD

**Issue by kyokley** [neovim/neovim#7383]
I know #7364 was recently merged in but my setup actually depends on cursorline taking precedence over syntax highlighting. I'm sure I'm most likely in the minority on this, but the ability to keep the current vim behavior would be extremely helpful. Do you think it would be possible to have a config setting that allows this?

**Comment by justinmk** [comment]
> my setup actually depends on cursorline taking precedence over syntax highlighting.

Can you explain how? Maybe with a screenshot of what it now looks like for you.

**Comment by kyokley** [comment]
In addition to using cursorline to highlight where the cursor is, I use it to display if I'm in insert mode or not. In normal mode, the cursorline is blue but I switch it to red when entering insert mode.

Nvim:
![nvim_normal](https://user-images.githubusercontent.com/11467843/31551172-5fdade22-aff9-11e7-971c-95f4576c0f4f.png)
![nvim_insert](https://user-images.githubusercontent.com/11467843/31551187-696b48e6-aff9-11e7-887a-fc79b5b075a6.png)

Vim:
![vim_normal](https://user-images.githubusercontent.com/11467843/31551204-72f7db40-aff9-11e7-81ab-80ff8d13a238.png)
![vim_insert](https://user-images.githubusercontent.com/11467843/31551211-78b2aa74-aff9-11e7-9074-afb43265a008.png)

**Comment by justinmk** [comment]
Ah, yes. If foreground is defined on CursorLine we can't just assume it should be low priority.

@zhou13 maybe a compromise would be:

- low-priority CursorLine if _only_ background is set (foreground is _not_ set)
- high-priority ("same as Vim" priority) CursorLine if foreground _and_ background are set

May need to revert https://github.com/neovim/neovim/commit/d1874ab2821d076397290cc154d87ec2dc352c79 for the time being.

**Comment by zhou13** [comment]
Yes, I think we can revert it for now.  The solution you proposed should work well.

**Comment by kyokley** [comment]
@justinmk @zhou13  Thanks guys

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/7400 reverts the regression.

Re-tagging this issue as "enhancement".

**Comment by indiesquidge** [comment]
> my setup actually depends on cursorline taking precedence over syntax highlighting.

My setup is also depending on this in order to "hide" my cursorline after 80 columns.

```vim
" Set 80 character ruler and colors
let &colorcolumn=join(range(81,999),",")
highlight ColorColumn cterm=NONE ctermbg=233

" Highlight current line and colors (must come last)
set cursorline
```

Before change (and currently on Vim 7.4):
<img width="2048" alt="screen shot 2017-11-15 at 1 25 00 pm" src="https://user-images.githubusercontent.com/3409645/32861058-712356cc-ca08-11e7-9483-87a482d66a5d.png">

Now:
<img width="2048" alt="screen shot 2017-11-15 at 1 24 44 pm" src="https://user-images.githubusercontent.com/3409645/32861065-74d5dfd8-ca08-11e7-929d-0aca745a9283.png">

**Comment by justinmk** [comment]
@indiesquidge Yep, we made a tradeoff. That use-case is far less common than what https://github.com/neovim/neovim/commit/7a832c312f13bd6be9350952cdac3bd30613a824 addresses, so this isn't going to change. 

Patches welcome if someone wants to find a way to address both uses-cases. https://github.com/neovim/neovim/commit/7a832c312f13bd6be9350952cdac3bd30613a824 won't be reverted.

**Comment by indiesquidge** [comment]
@justinmk, mmm I see. What was the use-case that 7a832c3 addresses? I'd love to help work on a solution that helps both settings coexist.

**Comment by justinmk** [comment]
@indiesquidge 7a832c3  addresses the problem where 'colorcolumn' hides text.

**Comment by iKarith** [comment]
Perhaps rather than assuming what the user intends, we should let them decide?  (I'm one of the folks who wants the priority lower unless I can specify "lighten" and "darken" as attributes for the cursorline foreground/backgrounds…)

**Comment by zhou13** [comment]
@indiesquidge In fact, before reverting 7a832c3 I guess your setup would work as `CursorLine`'s priority was lower than `ColorColumn`.  So your setup would work after this issue got solved.

**Comment by wincent** [comment]
FWIW I have a similar use case to @indiesquidge in that I don't necessarily want `ColorColumn` to be low-pri. I was using it to show long lines (longer than `'textwidth'`).

This is how it looks in Vim (8):

![__vim___glh-mbp___dot___zsh__tmux_](https://user-images.githubusercontent.com/7074/34427495-b9c59e10-ebf7-11e7-89e2-d1f545cc9387.png)

This is how it looks in Neovim (v0.2.2):

![nvim___glh-mbp___dot___zsh__tmux_](https://user-images.githubusercontent.com/7074/34427511-e1ad72e0-ebf7-11e7-8fb5-1bf132803a07.png)

`ColorColumn` is bg-only (ie. `ColorColumn    xxx ctermbg=10 guibg=#282a2e`), and in Neovim you can see the function names indicated by the arrows end up overwriting that background because they are highlighted with `vimUserFunc` which links to `Normal`:

```
:hi vimUserFunc
vimUserFunc    xxx links to Normal
:hi Normal
Normal         xxx ctermfg=7 ctermbg=0 guifg=#c5c8c6 guibg=#1d1f21
```

Other items such as the `setlocal` keywords link to `Statement      xxx ctermfg=1 gui=bold guifg=#cc6666` which doesn't mess with the background, so they look ok.

Now, I can "fix" this by patching any color scheme that creates backgrounds that I don't want. For example, I can just do `:hi clear Normal` and things look ending up reasonable, but there are probably a lot of color schemes where that kind of workaround might be difficult or may not work at all.

This is definitely a tricky one to balance.

**Comment by zhou13** [comment]
@wincent Thanks for posting your finding!  I think you can fix it by using
````
:hi vimUserFunc link NONE
````

The problem is the `hi def link vimUserFunc       Normal` in `runtime/syntax/vim.vim`.  IMO, when you need to link something to `Normal`, you really want it to be normal, even under `cursorcolumn/cursorline/colorcolumn`. I am not sure why the original author of `vim.vim` wants to bind `vimUserFunc` to `Normal` in the first place.  Maybe we just need to change that?

**Comment by zhou13** [comment]
However after searching `runtime/syntax/*.vim`, I found that there are a lot of syntax files that try to bind things to `Normal`.  I guess we need to find out a way to solve the problem?  Otherwise, after lowing the priority of `cursorline`, such bug can become obvious.  One workaround may just ignore the link to `Normal`.

**Comment by mox-mox** [comment]
Just a stupid question, but how difficult would it be to add a setting for cursorline priority?
I noticed that cursorline more or less just vanishes when the cursor is over a fold. If there was a way to specify the priority of either CursorLine or Folded highlight groups this could be prevented.

**Comment by zhou13** [comment]
Under current framework, changing anything related to highlight is difficult.  You need to go through a function called `win_line`, which contains thousands of lines of code that highlight different stuff in some brittle way. So we need to be extremely careful not to mess up any existing user cases....

**Comment by bfredl** [comment]
The highlighting system could in principle be refactored to add priority as an explicit attribute. Already #8221 will allow external uis to "reinterpret" combined highlight and thus change prioritization, though a native system/interface would be preferable in the long run.

**Comment by justinmk** [comment]
#8578 addresses the [original request](https://github.com/neovim/neovim/issues/7383#issuecomment-336472619) here.

@indiesquidge After #8578, your [example](https://github.com/neovim/neovim/issues/7383#issuecomment-344733729) can be achieved if foreground is _not_ set on CursorLine, so it's lower-priority than ColorColumn:

    hi CursorLine ctermbg=darkred ctermfg=NONE

This is a compromise. Anything beyond that will entail implementing customization of hl-group priority as mentioned in https://github.com/neovim/neovim/issues/7383#issuecomment-383725504 .

Any further discussion should go in a new issue.

**Comment by blueyed** [comment]
For reference: https://github.com/neovim/neovim/pull/10256 brings the vim patch for `nocombine`, which could be used (at least) for `CursorLine` then I assume.
